### PR TITLE
Properties - cross compiled version

### PR DESCRIPTION
**Pull request by Starzu**
Changes: 
* `*Property` is compiled to both JS and JVM now
* standard class can be used as `ModelProperty` template 
* removed `ImmutableValue` -> `Property` can contain anything now
* `ModelPropertyCreator` has to be created explicitly (`extends HasModelPropertyCreator[T]` or `ModelPropertyCreator.materialize[T]`)
* `CallbackSequencer` is no longer an `object` - now it's `ThreadLocal`

### CODE REVIEW

**Comment by leonardehrenfried** [comment]
Does that mean that the immutability checks for view models are completely gone?

**Comment by Starzu** [comment]
@leonardehrenfried Yes.

File: core/macros/src/main/scala/io/udash/macros/PropertyMacros.scala [hunk 1]
```diff
@@ -71,187 +64,78 @@ class PropertyMacros(val ctx: blackbox.Context) extends AbstractMacroCommons(ctx
...
+  //Checks, if tpe is a class which can be used as ModelProperty template
+  private def doesMeetClassModelRequirements(tpe: Type): Boolean = {
+    tpe.typeSymbol.isClass && !tpe.typeSymbol.isAbstract &&
+      findPrimaryConstructor(tpe).paramLists.size == 1 &&
+      findPrimaryConstructor(tpe).paramLists(0).nonEmpty &&
+      findPrimaryConstructor(tpe).isPublic &&
+      filterMembers(tpe)
+        .filter(m => !m.isPrivate && !(tpe <:< typeOf[Tuple2[_, _]] && m.name.decodedName.toString == "swap"))
+        .forall { m =>
+          if (m.isMethod && m.asMethod.isAccessor && m.asMethod.accessed.isTerm) m.asMethod.accessed.asTerm.isVal
```
[the remark this answers is not in the document]
Comment (Starzu) [reply core/macros/src/main/scala/io/udash/macros/PropertyMacros.scala]: It can be `<none>` for a val defined in a trait.

File: build.sbt [hunk 2]
```diff
@@ -80,20 +94,28 @@ lazy val `core-macros` = project.in(file("core/macros"))
     libraryDependencies ++= Seq("org.scala-lang" % "scala-reflect" % scalaVersion.value)
   )
 
-lazy val `core-shared` = crossProject.crossType(CrossType.Pure).in(file("core/shared"))
-  .jsConfigure(_.dependsOn(`core-macros`))
-  .jvmConfigure(_.dependsOn(`core-macros`))
-  .settings(commonSettings: _*)
+lazy val `core-shared-JVM` = project.in(file("core/shared"))
```
Comment (ghik) [inline build.sbt]: This project's name is "core-shared-JVM" which is probably not what you want.

File: core/macros/src/main/scala/io/udash/macros/PropertyMacros.scala [hunk 3]
```diff
@@ -506,82 +318,58 @@ class PropertyMacros(val ctx: blackbox.Context) extends AbstractMacroCommons(ctx
...
+         |
+         |There are two valid model bases:
+         | * trait (not sealed trait) with following restrictions:
+         |   * it cannot contain vars
+         |   * it can contain implemented vals and defs, but they are not considered as subproperties
+         |   * all abstract vals and defs (without parameters) are considered as subproperties
+         | * (case) class with following restrictions:
+         |   * it cannot contain vars
+         |   * it can contain implemented vals and defs, but they are not considered as subproperties
+         |   * it cannot have more than one parameters list in primary constructor
```
[the remark this answers is not in the document]
Comment (Starzu) [reply core/macros/src/main/scala/io/udash/macros/PropertyMacros.scala]: I think it's a future work.

File: core/shared/src/main/scala/io/udash/properties/PropertyCreator.scala [hunk 4]
```diff
@@ -17,20 +17,13 @@ trait PropertyCreator[T] {
   }
 }
 
-object PropertyCreator {
-  def propertyCreator[T : PropertyCreator]: PropertyCreator[T] =
+object PropertyCreator extends PropertyCreatorImplicits {
+  trait MacroGeneratedPropertyCreator
```
Comment (ghik) [inline core/shared/src/main/scala/io/udash/properties/PropertyCreator.scala]: Pls comment why this was added for future readers

**Pull request merged**